json2junit/json2junit.go

```diff
@@ -0,0 +1,315 @@
+		return err
+	}
+	defer r.Close()
+
+	if err := os.MkdirAll(filepath.Dir(out), 0o755); err != nil {
+		return err
+	}
+
+	w, err := os.Create(out)
+	if err != nil {
+		return err
+	}
+
+	if err := Convert(w, r); err != nil {
+		w.Close()
+		return err
+	}
+	return w.Close()
+}
+
+// A Converter holds the state of a JSON-to-JUnit conversion.
+// It implements io.WriteCloser; the caller writes test output in,
+// and the converter writes JUnit output to w.
+//
+// The JSON input is buffered, so the caller can write it in arbitrary
+// size chunks that don't have to align with the JSON lines.
+// The JUnit output is not written immediately, but chunked into test
+// suites which are written to w as soon as they are complete.
+type Converter struct {
+	b      []byte // input buffer
+	suites []*junitTestSuite
+	w      io.Writer
+	xmlEnc *xml.Encoder
+}
+
+// NewConverter returns a "JSON to JUnit" converter.
+// Writes on the returned writer are written as JUnit to w,
+// with minimal delay.
+func NewConverter(w io.Writer) *Converter {
+	return &Converter{
+		w: w,
+	}
+}
+
+// Write writes a JSON test entry to the writer.
+func (c *Converter) Write(b []byte) (int, error) {
+	if c.xmlEnc == nil {
+		if err := c.openXML(); err != nil {
+			return 0, err
+		}
+	}
+	n := len(b)
+	for len(b) > 0 {
+		// Search for the next newline.
+		i := bytes.IndexByte(b, '\n')
+		if i < 0 {
+			// No newline, so just write the buffer.
+			c.b = append(c.b, b...)
+			break
+		}
+		data := b[:i]
+		if len(c.b) > 0 {
+			data = append(c.b, data...)
+			// Reset the buffer.
+			c.b = c.b[:0]
+		}
+
+		// Unmarshal the JSON.
+		// If the JSON is invalid, just ignore it.
+		var jsonEntry jsonEntry
+		if err := json.Unmarshal(data, &jsonEntry); err == nil {
+			// Process the JSON entry.
+			if err := c.processJSONEntry(jsonEntry); err != nil {
+				return 0, fmt.Errorf("failed to process line: %v", err)
+			}
+		}
+
+		// Skip the newline.
+		b = b[i+1:]
+	}
+	return n, nil
+}
+
+// Close marks the end of the go test output.
+func (c *Converter) Close() error {
+	if err := c.closeXML(); err != nil {
+		return fmt.Errorf("failed to close XML: %v", err)
+	}
+	return nil
+}
+
+// jsonEntry is a single entry in the JSON file.
+type jsonEntry struct {
+	Time    time.Time
+	Action  string
+	Package string
+	Test    string
+	Output  string
+	Elapsed float64
+}
+
+// The following structs definitions are taken from
+// https://llg.cubic.org/docs/junit/.
+
+type junitTestSuite struct {
+	XMLName   xml.Name   `xml:"testsuite"`
+	Name      string     `xml:"name,attr"`
+	Tests     int        `xml:"tests,attr"`
+	Failures  int        `xml:"failures,attr"`
+	Skipped   int        `xml:"skipped,attr"`
+	Time      float64    `xml:"time,attr"`
+	Timestamp string     `xml:"timestamp,attr"`
+	SystemOut *systemOut `xml:",omitempty"`
+	TestCases []*junitTestCase
+}
+
+func (s junitTestSuite) testCase(name string) *junitTestCase {
+	for _, tc := range s.TestCases {
+		if tc.Name == name {
```

Comment:
Hmm, it seems like `testCase` and `suite` are called in a decently tight loop, and I wonder if maps are actually still worth the trouble...

Reply:
Done 👍